File: src/services/fileImportService.ts
import { File } from 'expo-file-system';

const MAX_FILE_SIZE = 1_000_000; // 1MB
const VALID_EXTENSIONS = ['.txt', '.md', '.markdown'];

export interface FileImportResult {
  success: boolean;
  markdown?: string;
  fileName?: string;
  error?: string;
}

/**
 * Check if a URL is a file:// URL with a valid text/markdown extension.
 */
export function isFileImportUrl(url: string): boolean {
  if (!url.startsWith('file://')) return false;
  const path = decodeURIComponent(url.replace('file://', ''));
  const lower = path.toLowerCase();
  return VALID_EXTENSIONS.some((ext) => lower.endsWith(ext));
}

/**
 * Read a shared file from a file:// URL and return its content.
 */
export async function readSharedFile(url: string): Promise<FileImportResult> {
  try {
    const path = decodeURIComponent(url.replace('file://', ''));
    const fileName = path.split('/').pop() || 'unknown';

    if (!VALID_EXTENSIONS.some((ext) => path.toLowerCase().endsWith(ext))) {
      return { success: false, error: 'Unsupported file type. Only .txt, .md, and .markdown files are supported.' };
    }

    const file = new File(url);

    if (!file.exists) {
      return { success: false, error: 'File not found.' };
    }

    const size = file.size;
    if (size === 0) {
      return { success: false, error: 'File is empty.' };
    }
    if (size > MAX_FILE_SIZE) {
      return { success: false, error: 'File is too large (max 1MB).' };
    }

    const content = file.textSync();

    if (!content.trim()) {
      return { success: false, error: 'File is empty.' };
    }

    return { success: true, markdown: content, fileName };
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Failed to read file';
    return { success: false, error: message };
  }
}

File: native-ios/LiftMark/Services/FileImportService.swift
import Foundation

// MARK: - File Import Result

struct FileImportResult {
    let success: Bool
    let markdown: String?
    let fileName: String?
    let error: String?
}

// MARK: - FileImportService

enum FileImportService {

    private static let maxFileSize = 1_000_000 // 1 MB
    private static let validExtensions: Set<String> = ["txt", "md", "markdown"]

    // MARK: - URL Validation

    /// Check if a URL is a valid file import URL with a supported extension.
    static func isFileImportUrl(_ urlString: String) -> Bool {
        guard let fileUrl = toFileUrl(urlString) else { return false }
        let ext = (fileUrl.pathExtension).lowercased()
        return validExtensions.contains(ext)
    }

    // MARK: - Read Shared File

    /// Read the contents of a shared file from a file:// or liftmark:// URL.
    static func readSharedFile(_ urlString: String) -> FileImportResult {
        guard let fileUrl = toFileUrl(urlString) else {
            return FileImportResult(success: false, markdown: nil, fileName: nil, error: "Unsupported URL scheme.")
        }

        let path = fileUrl.path
        let fileName = fileUrl.lastPathComponent
        let ext = fileUrl.pathExtension.lowercased()

        // Validate extension
        guard validExtensions.contains(ext) else {
            return FileImportResult(
                success: false, markdown: nil, fileName: fileName,
                error: "Unsupported file type. Only .txt, .md, and .markdown files are supported."
            )
        }

        let fileManager = FileManager.default

        // Check file exists
        guard fileManager.fileExists(atPath: path) else {
            return FileImportResult(success: false, markdown: nil, fileName: fileName, error: "File not found.")
        }

        // Check file size
        do {
            let attributes = try fileManager.attributesOfItem(atPath: path)
            let fileSize = attributes[.size] as? Int ?? 0

            if fileSize == 0 {
                return FileImportResult(success: false, markdown: nil, fileName: fileName, error: "File is empty.")
            }
            if fileSize > maxFileSize {
                return FileImportResult(success: false, markdown: nil, fileName: fileName, error: "File is too large (max 1MB).")
            }
        } catch {
            return FileImportResult(success: false, markdown: nil, fileName: fileName, error: "Failed to check file size: \(error.localizedDescription)")
        }

        // Read file content
        do {
            let content = try String(contentsOfFile: path, encoding: .utf8)

            guard !content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                return FileImportResult(success: false, markdown: nil, fileName: fileName, error: "File is empty.")
            }

            return FileImportResult(success: true, markdown: content, fileName: fileName, error: nil)
        } catch {
            return FileImportResult(success: false, markdown: nil, fileName: fileName, error: "Failed to read file: \(error.localizedDescription)")
        }
    }

    // MARK: - URL Conversion

    /// Convert a URL string to a file URL.
    /// Accepts file:// and liftmark:// schemes.
    private static func toFileUrl(_ urlString: String) -> URL? {
        if urlString.hasPrefix("file://") {
            return URL(string: urlString)
        }
        if urlString.hasPrefix("liftmark://") {
            let path = urlString.replacingOccurrences(of: "liftmark://", with: "")
            return URL(fileURLWithPath: "/\(path)")
        }
        return nil
    }
}
